### PR TITLE
Introduce MonoInnerProducerBase

### LINKED ISSUE #2616: Possible doOnDiscard leak with `Flux.singleOrEmpty`
The `DoOnDiscardShouldNotLeakTest` sometimes fails on the `singleOrEmpty` scenario.
Althought things have been improved in #2535, there are still intermittent failures with the following test:

```
ensureMultipleSubscribersSupportWithNoLeaksWhenRacingCancelAndOnNextAndRequest
```

### PR DESCRIPTION
**Pull request by ericbottard**
Introduce MonoInnerProducerBase as a tighter alternative to MonoSubscriber, with stronger encapsulation and stricter contract with
regard to its state machine. Derive MonoSingle's subscriber from it.

Fixes #2616

### CODE REVIEW

**Review by simonbasle — COMMENTED** [review]
some comments, need to review all this more in depth

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 1]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+		}
+
+		/**
+		 * Discard the given value, generally this.value field. Lets derived subscriber with further knowledge about
+		 * the possible types of the value discard such values in a specific way. Note that fields should generally be
+		 * nulled out along the discard call.
+		 *
+		 * @param v the value to discard
+		 */
+		protected final void discard(@Nullable O v) {
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2677]: Not sure I get your point !?
Will reword the javadoc though, pointing to `discardTheValue()`

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 2]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+							actual.onComplete();
+							return;
+						}
+
+					}
+				}
+			}
+		}
+
+		protected void doOnRequest(long n) {
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2726]: I'd rather wait till we rework other operators to see if that makes sense, wdyt?

This class works as-is (and is tested as such in its unit test). Those are indeed __additional__ hooks, not necessarily things that MUST be overridden.

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 3]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+				}
+				if (STATE.compareAndSet(this, s, s|HAS_VALUE)) {
+					return;
+				}
+			}
+		}
+
+		@Override
+		public final void cancel() {
+			doOnCancel();
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2753]: Thought about that. Would need to apply to all hooks for consistency IMO then. Again, would like to see if some operators need to be notified in all cases to decide.

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2678]: the javadoc seems to imply sub-classes would need to override this method to fine-tune the way their specific values are discarded. but the method is `final`, so sub-classes won't be able to override it.

File: reactor-core/src/main/java/reactor/core/publisher/MonoSingle.java [hunk 4]
```diff
@@ -169,13 +166,13 @@ public void onComplete() {
...
-					actual.onError(Operators.onOperatorError(this,
+					actual().onError(Operators.onOperatorError(this,
 							new NoSuchElementException("Source was empty"),
-							actual.currentContext()));
+							actual().currentContext()));
 				}
 			}
 			else if (c == 1) {
-				complete(this.value);
+				complete();
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoSingle.java:175]: I wonder if having both `complete(T)` and `complete()` isn't going to introduce more confusion. Maybe we can find better naming for both? The `complete()` method doesn't have any javadoc, which doesn't help.

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2727]: right, it isn't public yet (which I initially missed) :+1:

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2754]: that is acceptable, yeah

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 5]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+		/**
+		 * Set the value internally, without impacting request tracking state.
+		 * This however discards the provided value when detecting a cancellation.
+		 *
+		 * @param value the new value.
+		 * @see #complete(Object)
+		 */
+		protected final void setValue(@Nullable O value) {
+			this.value = value;
+			while(true) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:2740]: let's be consistent and use `for(;;)` for infinite loops

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 6]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+		/**
+		 * Tries to emit the value and complete the underlying subscriber or
+		 * stores the value away until there is a request for it.
+		 * <p>
+		 * Make sure this method is called at most once
+		 * @param v the value to emit
+		 */
+		public final void complete(O v) {
+			doOnComplete(v);
+			while(true) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:2615]: let's be consistent and use `for(;;)` for infinite loops

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 7]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+
+		}
+
+		protected void doOnComplete() {
+
+		}
+
+		public final void complete() {
+			doOnComplete();
+			while(true) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:2646]: let's be consistent and use `for(;;)` for infinite loops

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 8]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+		public final boolean isCancelled() {
+			return state == CANCELLED;
+		}
+
+
+		@Override
+		public void request(long n) {
+			if (validate(n)) {
+				doOnRequest(n);
+				while (true) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:2705]: let's be consistent and use `for(;;)` for infinite loops

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 9]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+	/*package*/ static class MonoInnerProducerBase<O>
+			implements InnerProducer<O> {
+
+		private final CoreSubscriber<? super O> actual;
+
+		/**
+		 * The value stored by this Mono operator.
+		 */
+		@Nullable
+		private volatile O   value;
```
Comment (OlegDokuka) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:2586]: value can be non-volatile

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 10]
```diff
@@ -2574,6 +2574,205 @@ private LiftFunction(@Nullable Predicate<Publisher> filter,
...
+		public final void complete(O v) {
+			doOnComplete(v);
+			while(true) {
+				int s = this.state;
+				if (s == CANCELLED) {
+					discard(v);
+					return;
+				}
+
+				if ((s & HAS_REQUEST) == HAS_REQUEST && STATE.compareAndSet(this, s, s | (HAS_VALUE | HAS_COMPLETED))) {
```
Comment (OlegDokuka) [inline reactor-core/src/main/java/reactor/core/publisher/Operators.java:2622]: let's create static methods such as `boolean hasRequest(long state)` and of others bit check so the code becomes much more readable (at least that what I usually do)

**Review by OlegDokuka — CHANGES_REQUESTED** [review]

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java (hunk 9 above)
Comment (ericbottard) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2586]: Can it really ? Does a CAS on `state` create a barrier that affects every field?

**Review by simonbasle — APPROVED** [review]
it looks like a pretty good start for the MonoSubscriber overhaul to me.
@ericbottard and @OlegDokuka I'll let you finalize things and do the necessary polishing, I'm confident this has advanced enough that it can be merged.

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java (hunk 9 above)
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java:2586]: Yeah. It does, value is written before write barriers. Also in java (compares to cpp) when we do compareAndSet we use [`memory_order_acq_rel`](https://stackoverflow.com/questions/7363462/what-are-the-similarities-between-the-java-memory-model-and-the-c11-memory-mod) type of barrier which guarantees that all writes will not be reordered with that barrier and all subsequent read happened with `memory_order_acquire` will see nonvolatile changes happened before ``memory_order_acq_rel` barrier (see for more info -> https://en.cppreference.com/w/cpp/atomic/memory_order)

**Review by OlegDokuka — APPROVED** [review]
LGTM

**Pull request merged**